controller/instance_manager_controller.go

```diff
@@ -859,26 +867,46 @@ func (w *ReplicaManagerWatch) StopWatch() {
 	close(w.stopCh)
 }
 
-func engineToProcessStatus(engine *api.Engine) types.InstanceProcessStatus {
-	return types.InstanceProcessStatus{
-		Endpoint:  engine.Endpoint,
-		ErrorMsg:  engine.ProcessStatus.ErrorMsg,
-		Listen:    engine.Listen,
-		Name:      engine.Name,
-		PortEnd:   engine.ProcessStatus.PortEnd,
-		PortStart: engine.ProcessStatus.PortStart,
-		State:     types.InstanceState(engine.ProcessStatus.State),
-		Type:      types.InstanceTypeEngine,
+func updateInstancesForInstanceManager(im *longhorn.InstanceManager, newInstance *types.InstanceProcess, deleted bool) {
+	name := newInstance.Spec.Name
+	currentInstance, exist := im.Status.Instances[name]
+	newInstance.Spec = currentInstance.Spec
+	if !exist {
+		logrus.Warnf("Cannot find instance %v in instance manager %v", name, im.Name)
+		return
+	}
+	if currentInstance.Status.ResourceVersion >= newInstance.Status.ResourceVersion {
+		logrus.Debugf("Instance manager %v will ignore expired instance process %v", im.Name, name)
+		return
+	}
+
+	if deleted {
+		// The instance process shouldn't become state `deleted` without the related `DeletedAt` set.
+		// But the following race condition may cause the case:
+		//     1. Instance manager controller (imc) watch func gets latest result: instance becomes `deleted`  -->  the related entry will be removed
+		//     2. Instance handler (ih) creates a new instance with the same name  -->  ih will add entry for instance manager
+		//     3. imc watch func gets latest result: instance becomes `starting`  -->  the entry in the instance manager will be updated
+		//     4. imc poll func gets expired result: instance becomes `deleted` (here `DeletedAt` is empty)  --> should ignore this result/case
+		// Hence Ignoring this case can solve this race condition.
+		if currentInstance.Spec.DeletedAt != "" {
+			delete(im.Status.Instances, name)
+		}
+		return
 	}
-}
 
-func processToProcessStatus(process *api.Process) types.InstanceProcessStatus {
-	return types.InstanceProcessStatus{
-		ErrorMsg:  process.ProcessStatus.ErrorMsg,
-		Name:      process.Name,
-		PortEnd:   process.ProcessStatus.PortEnd,
-		PortStart: process.ProcessStatus.PortStart,
-		State:     types.InstanceState(process.ProcessStatus.State),
-		Type:      types.InstanceTypeReplica,
+	switch newInstance.Status.State {
+	case types.InstanceStateStopping:
+		fallthrough
+	case types.InstanceStateStopped:
+		if currentInstance.Spec.DeletedAt == "" {
```

Comment:
Should I remove this?

Reply:
Yeah, better remove this, since the unexpected stop should trigger the logic in the instance handler anyway.